### PR TITLE
Add support for clipping/cutting out (#435)

### PR DESCRIPTION
**Pull request by kleisauke**
Usage: `overlayWith('overlayimage.png', { cutout: true } )`

Only the API Docs and changelog needs to be updated. Test cases are included.

Thanks to @jcupitt for the help provided [here](https://github.com/jcupitt/libvips/issues/59#issuecomment-222351004).

### CODE REVIEW

File: src/operations.cc [hunk 1]
```diff
@@ -82,6 +82,70 @@ namespace sharp {
...
   /*
+    Cutout src over dst with given gravity.
+   */
+  VImage Cutout(VImage mask, VImage dst, const int gravity) {
+    using sharp::CalculateCrop;
+    using sharp::HasAlpha;
+    using sharp::MaximumImageAlpha;
+
+    bool maskHasAlpha = HasAlpha(mask);
+    bool dstHasAlpha = HasAlpha(dst);
```
Comment (lovell) [inline src/operations.cc]: I think `dstHasAlpha` will always be `true` as there's logic elsewhere to ensure the presence of an alpha channel.

File: src/operations.cc [hunk 2]
```diff
@@ -82,6 +82,70 @@ namespace sharp {
...
+
+    if (!maskHasAlpha && mask.bands() > 1) {
+      throw VError("Image to be overlaid must have an alpha channel or one band");
```
Comment (lovell) [inline src/operations.cc]: Should this start "Overlay image must ..."?

File: src/common.cc [hunk 3]
```diff
@@ -277,4 +277,17 @@ namespace sharp {
     return std::make_tuple(left, top);
   }
 
+  /*
+    Return the image alpha maximum. Useful for combining alpha bands. scRGB
+    images are 0 - 1 for image data, but the alpha is 0 - 255.
+  */
+  int MaximumImageAlpha(VipsInterpretation interpretation) {
```
Comment (lovell) [inline src/common.cc]: Thanks for externalising this logic - something similar appears in quite a few places - I can now switch the other occurrences to use this.

File: src/pipeline.cc [hunk 4]
```diff
@@ -693,10 +694,15 @@ class PipelineWorker : public AsyncWorker {
           // the overlayGravity was used for extract_area, therefore set it back to its default value of 0
           baton->overlayGravity = 0;
         }
-        // Ensure overlay is premultiplied sRGB
-        overlayImage = overlayImage.colourspace(VIPS_INTERPRETATION_sRGB).premultiply();
-        // Composite images with given gravity
-        image = Composite(overlayImage, image, baton->overlayGravity);
+        if(baton->overlayCutout) {
+          // 'cut out' the image, premultiplication is not required
```
Comment (lovell) [inline src/pipeline.cc]: The image to be overlaid has been pre-multiplied, so we'll either need to also pre-multiply here or not pre-multiply the image to be overlaid via the logic that calculates `shouldPremultiplyAlpha`.

**Comment by lovell** [comment]
Fantastic, thank you for this Kleis. It's always a pleasure to see tests and I'm happy to update the docs etc. I've added a few comments inline.

File: src/operations.cc (hunk 1 above)
Comment (kleisauke) [inline src/operations.cc]: You're right. https://github.com/lovell/sharp/blob/master/src/pipeline.cc#L448-L455

File: src/operations.cc (hunk 2 above)
Comment (kleisauke) [inline src/operations.cc]: I'll fix that.

File: src/common.cc (hunk 3 above)
Comment (kleisauke) [inline src/common.cc]: :thumbsup:

File: src/pipeline.cc (hunk 4 above)
Comment (kleisauke) [inline src/pipeline.cc]: What will be the best option? Will there be any difference in the image if we don't pre-multiply it?

**Comment by lovell** [comment]
Updates look great, thank you.

> "What will be the best option? Will there be any difference in the image if we don't pre-multiply it?"

I think the best approach is to avoid premultiplying the image to be overlaid, assuming there are no other reasons for `shouldPremultiplyAlpha` being set to `true`, when using `cutout`.

This will avoid (a) a performance hit, (b) a source of rounding/clipping errors, and (c) potential light/dark artefacts at the "edges" of what was the discarded alpha channel.

**Comment by jcupitt** [comment]
A general Porter-Duff compose operator could solve this problem too. vips will need PD at some point.

**Comment by lovell** [comment]
@kleisauke Would you like to remove the premultiply logic here or would you prefer I take a look?

**Comment by kleisauke** [comment]
@lovell I already did that. See: [pipeline.cc#L463-L464](https://github.com/kleisauke/sharp/blob/feature-cutout/src/pipeline.cc#L463-L464) and [pipeline.cc#L698-L700](https://github.com/kleisauke/sharp/blob/feature-cutout/src/pipeline.cc#L698-L700).

**Comment by lovell** [comment]
@kleisauke You are absolutely correct - sorry I completely missed that. Thank you again for adding this very useful feature!

**Comment by kleisauke** [comment]
@lovell No problem! Thanks for merging the pull request! :thumbsup:

**Pull request merged**